### ISSUE TITLE
Question about bingen bindings

### ISSUE THREAD

**Issue by QuantumEF** [ryankurte/rust-bladerf#4]
It looks like you all are using your own version of rust-bindgen generated bindings of libbladerf.
I wanted to ask if yall were aware of the official bindings at : https://github.com/Nuand/libbladeRF-rust ? It seems relatively newer compared to the age of this project, so I assumed you might not know about it yet. 

Would it be worthwhile moving over to the official bindings?
Further, are there any issues with the project I could help with?
I am interested in this project and at some point would like to make a wrapper using [seify](https://crates.io/crates/seify) for use with [FutureSDR](https://www.futuresdr.org/).

**Comment by ryankurte** [comment]
yeah i started working on these a loong time ago (maybe my first foray into rust) and both the ecosystem and i have changed a bit since then! no opposition to adopting their bindings, but, also not sure it offers any particular advantage, and their repo seems like a quick first effort rather than a more systematic approach.

if the goal is to improve the state of bladerf bindings i think the best things to do would be:

- [x] swap to compile-time bindgen now it's not particularly arduous to run
- [x] separate this into a [bladerf-sys crate](https://doc.rust-lang.org/cargo/reference/build-scripts.html#-sys-packages) containing the bindings  the `bladerf` crate that wraps this to expose safe interfaces
  - [x] sort out docs build for `bladerf-sys` (https://github.com/rust-lang/crates-build-env/pull/156)
- [ ] fix up / complete wrapper crate
- [ ] add github actions CI (and maybe [nextest](https://nexte.st/)) for non-hardware tests
- [ ] add/move integration tests to a separate crate so these can be run when hardware is available
- [ ] implement [seify::DeviceTrait](https://docs.rs/seify/0.12.0/seify/trait.DeviceTrait.html)
- [ ] write README's
- [ ] add examples?

which should all be pretty straightforward, i've swapped to bindgen and added the `-sys` split as a start, but there's lots of changes to the library as well as missing wrappers that i don't have time to look at at the moment if you're interested in continuing.

**Comment by QuantumEF** [comment]
Thanks for your reply. When I initially was looking at this repo, it seemed like it was being done a different way. Now that you are also using the autogenerated bindgen bindings, I don't see to much of a difference.

As I have time, I am going to continue looking into the code to better understand how libbladerf itself works, and hopefully actually contribute to this project. We will see though. 

Feel free to close the issue, though I think you should leave it open since it provides a useful todo list.

**Comment by ryankurte** [comment]
>  When I initially was looking at this repo, it seemed like it was being done a different way.

yeah before it was a combination of old generated bindings mixed together with custom types / and safe wrappers, not great practice by modern c library binding standards ^_^

> Now that you are also using the autogenerated bindgen bindings, I don't see to much of a difference.

the other reason to keep -sys bindings here is to be able to polyfill methods etc. on generated types without running afoul of the orphan rule (though you can mitigate this to some extent with traits), so i think it's best to keep them together.

> As I have time, I am going to continue looking into the code to better understand how libbladerf itself works, and hopefully actually contribute to this project. We will see though.

no stress either way, let me know if i can clarify anything ^_^

**Comment by QuantumEF** [comment]
I am probably going to move to contributing to this implementation of the bladerf bindings (https://github.com/MerchGuardian/seify-bladerf) since I like how they have done a lot more work with the types.

We have some discussion going on in the FutureSDR discord: https://discord.com/invite/vCz29eDbGP/
